packages/tools/src/eventListeners/touch/touchStartListener.ts

```diff
@@ -290,126 +335,128 @@ function _onTouchDrag(evt: TouchEvent) {
+  if (currentTime - tapState.startTime.getTime() >= tapState.tapWindow) return;
+
+  // subsequent tap is on a different element
   if (
-    state.taps > 0 &&
-    getDeltaDistance(currentPointsList, lastPointsList).canvas <
-      state.tapMaxDistance
+    tapState.taps > 0 &&
+    !(
+      tapState.element == state.element &&
+      tapState.renderingEngineId == state.renderingEngineId &&
+      tapState.viewportId == state.viewportId
+    )
   ) {
-    state.taps = state.taps + 1;
+    return;
   }
 
-  if (state.taps === 0) {
-    state.taps = state.taps + 1;
-  }
+  const currentPointsList = getTouchEventPoints(evt, tapState.element);
+  const distanceFromStart = getDeltaDistance(
+    currentPointsList,
+    tapState.startPointsList
+  ).canvas;
 
-  state.lastPointsList = copyPointsList(currentPointsList); // Update the last points
+  if (distanceFromStart < tapState.tapMaxDistance) {
+    clearTimeout(tapState.tapTimeout);
+    tapState.startTime = new Date();
+    tapState.taps += 1;
+  }
 
-  state.tapTimeout = setTimeout(() => {
+  tapState.tapTimeout = setTimeout(() => {
     const eventDetail: EventTypes.TouchTapEventDetail = {
       event: evt,
       eventName: TOUCH_TAP,
-      element: state.element,
-      renderingEngineId: state.renderingEngineId,
-      viewportId: state.viewportId,
+      element: tapState.element,
+      renderingEngineId: tapState.renderingEngineId,
+      viewportId: tapState.viewportId,
       camera: {},
       currentPointsList,
       currentPoints: getMeanTouchPoints(currentPointsList),
-      taps: state.taps,
+      taps: tapState.taps,
     };
     triggerEventCallback(eventDetail.element, TOUCH_TAP, eventDetail);
-    state = JSON.parse(JSON.stringify(defaultState));
-  }, state.touchStartDelay - 10);
+    tapState = JSON.parse(JSON.stringify(defaultTapState));
+  }, tapState.tapWindow);
 }
 
 function _checkTouchSwipe(evt: TouchEvent, deltaPoints: IPoints) {
+  const currentTime = new Date().getTime();
+  const startTime = state.startTime.getTime();
+  if (state.swiped || currentTime - startTime >= state.swipeWindow) return;
   const [x, y] = deltaPoints.canvas;
-  if (
-    Math.abs(x) > state.swipeDistanceThreshold ||
-    Math.abs(y) > state.swipeDistanceThreshold
-  ) {
-    const eventDetail: EventTypes.TouchSwipeEventDetail = {
-      event: evt,
-      eventName: TOUCH_SWIPE,
-      renderingEngineId: state.renderingEngineId,
-      viewportId: state.viewportId,
-      camera: {},
-      element: state.element,
-      swipe: null,
-    };
+  const eventDetail: EventTypes.TouchSwipeEventDetail = {
+    event: evt,
+    eventName: TOUCH_SWIPE,
+    renderingEngineId: state.renderingEngineId,
+    viewportId: state.viewportId,
+    camera: {},
+    element: state.element,
+    swipe: null,
+  };
+  if (Math.abs(x) > state.swipeDistanceThreshold) {
+    eventDetail.swipe = x > 0 ? Swipe.RIGHT : Swipe.LEFT;
+    triggerEventCallback(eventDetail.element, TOUCH_SWIPE, eventDetail);
```

Comment:
That's right, previously TOUCH_SWIPE and TOUCH_DRAG were mutually exclusive; however, this implementation required a setTimeout delay which makes either drag or swipe feel laggy. (Currently this is the case with Tap Event, see comment regarding whether to push down to embedding application layer)

Application layer will need to manage SWIPE/DRAG events happening together to have responsive touch.

For most use cases, drag and swipe gestures don't really conflict. And since drag event is clamped to 10 the affect is not that bad for a mis swipe or mis drag.

Reply:
Having them mutually exclusive is fine - leave this as is.